File: app/src/main/java/mariavv/fitnesspal/presentation/NavigatePresenter.java
package mariavv.fitnesspal.presentation;

import android.annotation.SuppressLint;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import mariavv.fitnesspal.FitnessPal;
import mariavv.fitnesspal.data.repository.AssetsRepository;
import mariavv.fitnesspal.data.repository.Repo;
import mariavv.fitnesspal.data.repository.SharedDataRepository;
import mariavv.fitnesspal.domain.Dish;
import mariavv.fitnesspal.domain.Dish2;
import mariavv.fitnesspal.domain.Food;
import mariavv.fitnesspal.domain.TestData;
import mariavv.fitnesspal.other.Const;
import mariavv.fitnesspal.other.FrmFabric;
import ru.terrakok.cicerone.Router;

@InjectViewState
public class NavigatePresenter extends MvpPresenter<NavigateView> {

    private Router router;

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        router = FitnessPal.instance.getRouter();

        if (!SharedDataRepository.isNotFirstRun()) {
            initDb();
            SharedDataRepository.saveNotFirstRun(true);
        } else {
            router.newRootScreen(Const.Screen.JOURNAL);
        }
    }

    void onNavigationJournalSelected() {
        router.newRootScreen(Const.Screen.JOURNAL);
    }

    void onNavigationHandbookSelected() {
        router.newRootScreen(Const.Screen.HANDBOOK);
    }

    void onResume() {
        getViewState().setNavigator();
    }

    void onPause() {
        getViewState().removeNavigator();
    }

    void changeFragment(FragmentManager fragmentManager, @IdRes int containerId) {
        final String screenName = getFragmentName(fragmentManager, containerId);
        final boolean needShowMenu = screenName.equals(Const.Screen.HANDBOOK)
                || screenName.equals(Const.Screen.JOURNAL);

        if (needShowMenu) {
            getViewState().showBottomMenu();
            getViewState().showHomeAsUp(false);
        } else {
            getViewState().hideBottomMenu();
            getViewState().showHomeAsUp(true);
        }
    }

    @NonNull
    private String getFragmentName(FragmentManager fragmentManager, @IdRes int containerId) {
        final Fragment fragment = fragmentManager.findFragmentById(containerId);
        if (fragment instanceof FrmFabric.IFragment) {
            return ((FrmFabric.IFragment) fragment).getName();
        }
        return "";
    }

    @SuppressLint("CheckResult")
    private void initDb() {
        //test data
        AssetsRepository repo = new AssetsRepository();

        //todo map
        Observable.fromCallable(repo.getTestData(FitnessPal.instance.getAssets()))
                .map(new Function<TestData, TestData>() {
                    @Override
                    public TestData apply(TestData testData) throws Exception {
                        return testData;
                    }
                }).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onGetTestData);
    }

    private void onGetTestData(TestData data) {
        final Repo repo = Repo.getInstance();

        for (Food food : data.foods) {
            repo.insertFoodInHandbook(food);
        }

        //todo dish2
        for (Dish2 dish : data.dishes) {
            //todo
            SimpleDateFormat format = new SimpleDateFormat();
            format.applyPattern("dd.MM.yyyy");
            Date docDate = new Date();
            try {
                docDate = format.parse(dish.date);
            } catch (ParseException e) {
                e.printStackTrace();
            }
            repo.insertDishInJournal(new Dish(docDate, dish.meal, dish.foodId, dish.weight));
        }

        router.newRootScreen(Const.Screen.JOURNAL);
    }
}

File: app/src/main/java/mariavv/fitnesspal/presentation/NavigatePresenter.kt
package mariavv.fitnesspal.presentation

import android.annotation.SuppressLint
import android.support.annotation.IdRes
import android.support.v4.app.FragmentManager
import android.util.Log
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers
import mariavv.fitnesspal.App
import mariavv.fitnesspal.R
import mariavv.fitnesspal.data.repository.AssetsRepository
import mariavv.fitnesspal.data.repository.DbRepository
import mariavv.fitnesspal.data.repository.SharedDataRepository
import mariavv.fitnesspal.domain.Dish
import mariavv.fitnesspal.domain.StartedData
import mariavv.fitnesspal.other.Const
import mariavv.fitnesspal.other.FrmFabric
import ru.terrakok.cicerone.Router
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

@InjectViewState
class NavigatePresenter : MvpPresenter<NavigateView>() {

    private lateinit var router: Router

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        router = App.getRouter()

        if (SharedDataRepository.isFirstRun) {
            initDb()
            SharedDataRepository.saveFirstRun()
        } else {
            router.newRootScreen(Const.Screen.JOURNAL)
        }
    }

    internal fun onNavigationJournalSelected() {
        router.newRootScreen(Const.Screen.JOURNAL)
    }

    internal fun onNavigationHandbookSelected() {
        router.newRootScreen(Const.Screen.HANDBOOK)
    }

    internal fun onResume() {
        viewState.setNavigator()
    }

    internal fun onPause() {
        viewState.removeNavigator()
    }

    internal fun changeFragment(fragmentManager: FragmentManager, @IdRes containerId: Int) {
        val screenName = getFragmentName(fragmentManager, containerId)
        val needShowMenu = screenName == Const.Screen.HANDBOOK || screenName == Const.Screen.JOURNAL

        if (needShowMenu) {
            viewState.showBottomMenu()
            viewState.showHomeAsUp(false)
        } else {
            viewState.hideBottomMenu()
            viewState.showHomeAsUp(true)
        }
    }

    private fun getFragmentName(fragmentManager: FragmentManager, @IdRes containerId: Int): String {
        val fragment = fragmentManager.findFragmentById(containerId)
        return if (fragment is FrmFabric.IFragment) {
            (fragment as FrmFabric.IFragment).name
        } else ""
    }

    @SuppressLint("CheckResult")
    private fun initDb() {
        //test data
        //todo map
        Observable.fromCallable(AssetsRepository().getTestData(App.context.assets))
                .map { testData -> testData }.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { this.onGetTestData(it) }
    }

    private fun onGetTestData(data: StartedData) {
        data.foods?.forEach { food ->
            Single.fromCallable {
                DbRepository.instance.insertFoodInHandbook(food)
            }.subscribeOn(
                    Schedulers.io()
            ).subscribeBy(
                    onError = { t ->
                        Log.d(Const.LOG_TAG, "test data insert error: ${t.message}")
                    },
                    onSuccess = { id ->
                        Log.d(Const.LOG_TAG, "test data insert success, id = $id")
                    }

            ).addTo(CompositeDisposable())
        }

        //todo dish2
        data.dishes?.forEach { dish ->
            //todo
            val format = SimpleDateFormat("dd.MM.yyyy", Locale.getDefault())
            var docDate = Date()
            try {
                docDate = format.parse(dish.date)
            } catch (e: ParseException) {
                e.printStackTrace()
            }

            Single.fromCallable {
                DbRepository.instance.insertDishInJournal(Dish(
                        meal = dish.meal, date = docDate.time, foodId = dish.foodId, weight = dish.weight
                ))
            }.subscribeOn(
                    Schedulers.io()
            ).observeOn(
                    AndroidSchedulers.mainThread()
            ).subscribeBy(
                    onSuccess = { id ->
                        if (id < 1) {
                            App.getRouter().showSystemMessage(App.getAppString(R.string.add_dish_fail))
                        }
                        Log.d(Const.LOG_TAG, "insert dish success, id = $id")
                    },
                    onError = { t ->
                        App.getRouter().showSystemMessage(App.getAppString(R.string.add_dish_fail))
                        Log.d(Const.LOG_TAG, "insert dish error: ${t.message}")
                    }
            ).addTo(
                    CompositeDisposable()
            )
        }

        router.newRootScreen(Const.Screen.JOURNAL)
    }
}
